### ISSUE TITLE
XGBoost 1.5.2 - Non-deterministic `XGBoostError` with `gpu_hist` for `ellpack_page.cu` on `NumSymbols()`

### ISSUE THREAD

**Issue by lebiathan** [dmlc/xgboost#9565]
Hi folks,

I'm doing single-GPU training w/ XGBoost 1.5.2 with a proprietary dataset & codebase.  I'll try to provide as much info as permissibly possible.

I am using the scikit-learn API to train the model & BayesianOptimization (https://github.com/bayesian-optimization/BayesianOptimization) for hyperparam selection.

The data used in training / validation is stored (on disk) in parquet and is read in memory via `pd.read_parquet()`, `pd` being an alias for `pandas`. This happens once at the very beginning. The machine has sufficient memory (RAM) to hold all the data at once.  There is no sampling involved prior to training.  All features are numerical (no categorical variables).  The shape of the dataset(s) (features only) is:
* training: [~7M, 180]
* validation: [~20M, 180]

I have a custom `xgboost.core.DataIter` instance (added in the end) as I was getting OOM errors without it. The training dataset is read in a single pass. The validation dataset requires 2 passes. There is no sampling involved here either.

On occasion, I get the following error early on in the training process:
```bash
  ... # truncated stacktrace that references internal code structure. Just reading data in memory and passing through the scikit API
  File "/mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/training.py", line 81, in _train_internal
    bst.update(dtrain, i, obj)
  File "/mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/core.py", line 1682, in update
    dtrain.handle))
  File "/mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/core.py", line 218, in _check_call
    raise XGBoostError(py_str(_LIB.XGBGetLastError()))
xgboost.core.XGBoostError: [13:39:34] ../src/tree/updater_gpu_hist.cu:770: Exception in gpu_hist: [13:39:34] ../src/data/ellpack_page.cu:375: Check failed: NumSymbols() == page->NumSymbols() (16647 vs. 16670) : 
Stack trace:
  [bt] (0) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x37b8a9) [0x7f32d76c08a9]
  [bt] (1) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x37d31d) [0x7f32d76c231d]
  [bt] (2) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x4a5ddd) [0x7f32d77eaddd]
  [bt] (3) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x4a3f19) [0x7f32d77e8f19]
  [bt] (4) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x4dee94) [0x7f32d7823e94]
  [bt] (5) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x4e1c88) [0x7f32d7826c88]
  [bt] (6) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x17d683) [0x7f32d74c2683]
  [bt] (7) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x17e207) [0x7f32d74c3207]
  [bt] (8) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x1b49fa) [0x7f32d74f99fa]



Stack trace:
  [bt] (0) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x4c2fe9) [0x7f32d7807fe9]
  [bt] (1) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x4e207f) [0x7f32d782707f]
  [bt] (2) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x17d683) [0x7f32d74c2683]
  [bt] (3) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x17e207) [0x7f32d74c3207]
  [bt] (4) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(+0x1b49fa) [0x7f32d74f99fa]
  [bt] (5) /mnt/task_runtime/venv/lib/python3.7/site-packages/xgboost/lib/libxgboost.so(XGBoosterUpdateOneIter+0x68) [0x7f32d73de4f8]
  [bt] (6) /miniconda/lib/python3.7/lib-dynload/../../libffi.so.7(+0x69dd) [0x7f339cff39dd]
  [bt] (7) /miniconda/lib/python3.7/lib-dynload/../../libffi.so.7(+0x6067) [0x7f339cff3067]
  [bt] (8) /miniconda/lib/python3.7/lib-dynload/_ctypes.cpython-37m-x86_64-linux-gnu.so(_ctypes_callproc+0x2e7) [0x7f339d00b517]
```

The error does not deterministically appear.  Using the same hyperparam (HP) search space in two different runs, one of them might fail, while the other one can go through. Even if I use specific hyperparams (no HP searching), the error does not always appear.   To be clear, the ratio `success:failure` is _not_ 1:1.

What _is_ deterministic about the error, it's always happened before the 1st estimator / booster is evaluated on the validation set, possibly earlier, e.g., before the first estimator has been fully built.  Once the 1st estimator is built & evaluated (read: I see logged metrics) for a specific set of HPs, I've never seen the error happen.

#### HP Params

The following hyperparams are just one example picked by bayesopt when the error occurred. As mentioned above, failures have happened for other hyperparams too. That said, I've tried separately with these exact HPs (no bayesopt) and the error did _not_ appear.

```python
hp_params = {
    "n_jobs": 10,
    "tree_method": "gpu_hist",
    "n_estimators": 650,
    "max_depth": 4,
    "learning_rate": 0.0202,
    "gamma": 0.3423,
    "subsample": 0.95
}
```

#### Virtual Env
Using Python 3.7.9
```python
bayesian-optimization==1.4.2
numpy==1.21.6
pandas==1.3.5
pyarrow==12.0.1
scipy==1.7.3
scikit-learn==1.0.2
xgboost==1.5.2
```

The custom DataIter looks like this (modified from https://xgboost.readthedocs.io/en/stable/tutorials/external_memory.html#data-iterator)
```python
from typing import Callable

import numpy as np
import pandas as pd
import xgboost.core


class CustomDataIter(xgboost.core.DataIter):
    def __init__(
            self, features: pd.DataFrame,  label: np.ndarray,  weight: np.ndarray,  batch_size: int,
    ):
        super().__init__()

        assert (label is None) or (len(features) == len(label))
        assert (weight is None) or (len(features) == len(weight))

        self.features = features
        self.labels = label
        self.weights = weight

        self.cur_offset = 0
        self.batch_size = batch_size
        return

    def reset(self):
        self.cur_offset = 0
        return

    def next(self, input_data: Callable) -> int:
        if self.cur_offset >= len(self.features):
            return 0

        end_offset = min(len(self.features), self.cur_offset + self.batch_size)

        features_subset = self.features[self.cur_offset:end_offset]
        labels_subset = self.labels[self.cur_offset:end_offset] if self.labels is not None else None
        weights_subset = self.weights[self.cur_offset:end_offset] if self.weights is not None else None

        input_data(data=features_subset, label=labels_subset, weight=weights_subset)

        self.cur_offset = end_offset
        return 1
```

Thanks in advance for your help!

**Comment by trivialfis** [comment]
Could you please share how you pass the data iter into xgboost?

**Comment by lebiathan** [comment]
Hi @trivialfis ,

Here's an abridged (but accurate) version of how we pass the data to XGBoost

```python
from typing import List, Tuple
import numpy as np
import pandas as pd

def read_data(dirname: str, feature_columns: List[str]) -> Tuple[pd.DataFrame, np.ndarray]:
    df = pd.read_parquet(dirname)
    return df[feature_columns], df["label"].values


# This will read everything in memory first
X_train, y_train = read_data("/path/to/data/training/", features)
train_iter = CustomDataIter(X_train, y_train, None, BATCH_SIZE)
dtrain = xgboost.DMatrix(train_iter)  # following the example in: https://xgboost.readthedocs.io/en/stable/tutorials/external_memory.html#data-iterator

X_validation, y_validation = read_data("/path/to/data/validation/", features)
validation_iter = CustomDataIter(X_validation, y_validation, None, BATCH_SIZE)
dvalidation = xgboost.DMatrix(validation_iter)

evals_result = {}
bst = xgboost.train(
    params,  # contains 'eval_metric': ['logloss', 'auc'].  See earlier comment for other params
    dtrain,
    num_boost_round=n_estimators,
    evals=[(dtrain, 'train'), (dvalidation, 'validation')], 
    evals_result=evals_result,
)
```

**Comment by trivialfis** [comment]
Hmm, I can't reproduce the error. Could you please try the latest version of XGBoost?

**Comment by lebiathan** [comment]
The error does not happen with `xgboost==1.7.6`. I am getting a segmentation fault immediately with version `1.6.2` with the same code I showed above, even with very small datasets we use for basic testing (< 2K rows, ~100 features), without trying to split the dataset (i.e., the `BATCH_SIZE` is big enough to return the dataset at once). If I remove the custom iterator in 1.6.2, everything works fine (with the small basic testing dataset).  We do not have custom low-level code in our repo for the seg fault to be coming from that.

I have not tried v2 that just came out.

When the error occurs, I do see some `DMatrix...`-prefixed files, which I suppose contain debug info. Would it be possible to share these privately -- pending security / manager approvals -- in case they help figure things out?

Note: I understand that upgrading the library addresses this problem but a) I've noticed some decrease in training performance (speed) with `1.7.6` compared to `1.5.2` (separate issue) and b) that would require a few broader changes on our end.

Thanks again!

**Comment by trivialfis** [comment]
Hmm, you are using external memory... It's still considered experimental, and we have made some huge optimization in 2.0 along with comprehensive documentation on best practice, you might want to try it.

Alternatively, you can try the `QuantileDMatrix` with `hist` for saving memory. It reduces memory usage without loss of performance. (to me, performance of external memory was less than bearable before 2.0)

I don't think we will continue to backport to previous releases older than 1.7.x, I (as an individual instead of a maintainer, the XGBoost team doesn't make such promise) can help with 1.7.x since 2.0 is kind of a big change and I expect some people want to stay with 1.7.x for a period of time.

> I've noticed some decrease in training performance (speed) with 1.7.6 compared to 1.5.2 (separate issue) 

Maybe try the `hist` tree method?

**Comment by lebiathan** [comment]
Thanks @trivialfis . It's totally understandable that support for older versions will not be a priority moving forward.

I misunderstood the comment from here: https://github.com/dmlc/xgboost/releases/tag/v1.6.0 
> External memory support for both approx and hist tree method is considered feature complete in XGBoost 1.6.

to mean that the iterator would work well for `gpu_hist` as well.

In any case, if you think that sharing the core dump files I mentioned above would help uncover & improve something in later versions, let me know and we can see how to make those files available.  Thanks!

**TL;DR**
For future travelers who may experience the same issue: Upgrade to `1.7.6` or later.

**Comment by trivialfis** [comment]
> I misunderstood the comment from here

Apologies for that comment, it's still an experimental feature. Has been going through a lot of changes since then and a lot more is needed to make it production-ready.

> to mean that the iterator would work well for gpu_hist as well.

We weren't aware of the bug in 1.5.x. But back then, the feature was still quite experimental, so I wouldn't be suprised.

> In any case, if you think that sharing the core dump files I mentioned above would help uncover & improve something in later versions,

Thank you for sharing, I think that's unnecessary as external memory has been changed quite significantly in 2.0.
